File: bluecross/protocol.py
"""Network protocol for BlueCross communication."""

import asyncio
import json
import struct
from dataclasses import dataclass
from enum import IntEnum
from typing import Optional


class MessageType(IntEnum):
    """Types of messages sent over the network."""
    HANDSHAKE = 1
    HANDSHAKE_ACK = 2
    KEY_EVENT = 3
    MOUSE_MOVE = 4
    MOUSE_BUTTON = 5
    MOUSE_SCROLL = 6
    SWITCH_TO_CLIENT = 7
    SWITCH_TO_SERVER = 8
    HEARTBEAT = 9
    CLIPBOARD_DATA = 10


@dataclass
class Message:
    """A network message."""
    msg_type: MessageType
    payload: dict

    def encode(self) -> bytes:
        """Encode message to bytes."""
        payload_bytes = json.dumps(self.payload).encode("utf-8")
        header = struct.pack("!BH", self.msg_type, len(payload_bytes))
        return header + payload_bytes

    @classmethod
    def decode(cls, data: bytes) -> tuple["Message", bytes]:
        """Decode message from bytes, return message and remaining data."""
        if len(data) < 3:
            raise ValueError("Incomplete header")
        
        msg_type, payload_len = struct.unpack("!BH", data[:3])
        
        if len(data) < 3 + payload_len:
            raise ValueError("Incomplete payload")
        
        payload_bytes = data[3:3 + payload_len]
        payload = json.loads(payload_bytes.decode("utf-8"))
        
        return cls(MessageType(msg_type), payload), data[3 + payload_len:]


class ProtocolHandler:
    """Handles protocol communication over a stream."""
    
    def __init__(
        self,
        reader: asyncio.StreamReader,
        writer: asyncio.StreamWriter,
    ):
        self.reader = reader
        self.writer = writer
        self._buffer = b""
    
    async def send(self, msg: Message) -> None:
        """Send a message."""
        self.writer.write(msg.encode())
        # Don't drain on every send - let TCP buffer handle it
    
    async def flush(self) -> None:
        """Flush pending writes."""
        await self.writer.drain()
    
    async def receive(self) -> Optional[Message]:
        """Receive a message."""
        while True:
            try:
                msg, self._buffer = Message.decode(self._buffer)
                return msg
            except ValueError:
                chunk = await self.reader.read(65536)
                if not chunk:
                    return None
                self._buffer += chunk
    
    def receive_all_buffered(self) -> list[Message]:
        """Receive all messages currently in the buffer without waiting."""
        messages = []
        while True:
            try:
                msg, self._buffer = Message.decode(self._buffer)
                messages.append(msg)
            except ValueError:
                break
        return messages
    
    def close(self) -> None:
        """Close the connection."""
        self.writer.close()


def create_handshake(client_name: str, screen_width: int, screen_height: int) -> Message:
    """Create a handshake message."""
    return Message(MessageType.HANDSHAKE, {
        "name": client_name,
        "screen_width": screen_width,
        "screen_height": screen_height,
    })


def create_handshake_ack(position: str, server_width: int, server_height: int) -> Message:
    """Create a handshake acknowledgment."""
    return Message(MessageType.HANDSHAKE_ACK, {
        "position": position,
        "server_width": server_width,
        "server_height": server_height,
    })


def create_key_event(code: int, value: int) -> Message:
    """Create a key event message."""
    return Message(MessageType.KEY_EVENT, {"code": code, "value": value})


def create_mouse_move(x: int, y: int, dx: int, dy: int) -> Message:
    """Create a mouse move message."""
    return Message(MessageType.MOUSE_MOVE, {"x": x, "y": y, "dx": dx, "dy": dy})


def create_mouse_button(button: int, value: int) -> Message:
    """Create a mouse button message."""
    return Message(MessageType.MOUSE_BUTTON, {"button": button, "value": value})


def create_mouse_scroll(dx: int, dy: int) -> Message:
    """Create a mouse scroll message."""
    return Message(MessageType.MOUSE_SCROLL, {"dx": dx, "dy": dy})


def create_switch_to_client(entry_x: int, entry_y: int) -> Message:
    """Create a switch to client message."""
    return Message(MessageType.SWITCH_TO_CLIENT, {"entry_x": entry_x, "entry_y": entry_y})


def create_switch_to_server() -> Message:
    """Create a switch to server message."""
    return Message(MessageType.SWITCH_TO_SERVER, {})


def create_clipboard_data(content: str) -> Message:
    """Create a clipboard data message."""
    return Message(MessageType.CLIPBOARD_DATA, {"content": content})

File: src/protocol.rs
use std::fmt;

use serde_json::Value;

use crate::secure::{SecureReader, SecureWriter};

/// Hard cap on a single message payload (decoded JSON). Bounds memory use when a
/// peer announces a length; also the ceiling for clipboard payloads.
pub const MAX_PAYLOAD: usize = 16 * 1024 * 1024;

/// Fixed message header size: 1 type byte + 4-byte big-endian length.
const HEADER_LEN: usize = 5;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u8)]
pub enum MessageType {
    Handshake = 1,
    HandshakeAck = 2,
    KeyEvent = 3,
    MouseMove = 4,
    MouseButton = 5,
    MouseScroll = 6,
    SwitchToClient = 7,
    SwitchToServer = 8,
    Heartbeat = 9,
    ClipboardData = 10,
}

impl MessageType {
    pub fn from_u8(v: u8) -> Option<Self> {
        match v {
            1 => Some(Self::Handshake),
            2 => Some(Self::HandshakeAck),
            3 => Some(Self::KeyEvent),
            4 => Some(Self::MouseMove),
            5 => Some(Self::MouseButton),
            6 => Some(Self::MouseScroll),
            7 => Some(Self::SwitchToClient),
            8 => Some(Self::SwitchToServer),
            9 => Some(Self::Heartbeat),
            10 => Some(Self::ClipboardData),
            _ => None,
        }
    }
}

impl fmt::Display for MessageType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{:?}", self)
    }
}

#[derive(Debug, Clone)]
pub struct Message {
    pub msg_type: MessageType,
    pub payload: Value,
}

impl Message {
    pub fn encode(&self) -> Vec<u8> {
        let payload_bytes =
            serde_json::to_vec(&self.payload).expect("payload is always serializable");
        let mut buf = Vec::with_capacity(HEADER_LEN + payload_bytes.len());
        buf.push(self.msg_type as u8);
        buf.extend_from_slice(&(payload_bytes.len() as u32).to_be_bytes());
        buf.extend_from_slice(&payload_bytes);
        buf
    }

    /// Decode a message from a buffer. Returns (message, bytes_consumed) or None if incomplete.
    pub fn decode(data: &[u8]) -> anyhow::Result<Option<(Self, usize)>> {
        if data.len() < HEADER_LEN {
            return Ok(None);
        }
        let msg_type_byte = data[0];
        let payload_len = u32::from_be_bytes([data[1], data[2], data[3], data[4]]) as usize;
        if payload_len > MAX_PAYLOAD {
            anyhow::bail!(
                "payload length {} exceeds maximum {}",
                payload_len,
                MAX_PAYLOAD
            );
        }
        if data.len() < HEADER_LEN + payload_len {
            return Ok(None);
        }
        let msg_type = MessageType::from_u8(msg_type_byte)
            .ok_or_else(|| anyhow::anyhow!("Invalid message type: {}", msg_type_byte))?;
        let payload: Value = serde_json::from_slice(&data[HEADER_LEN..HEADER_LEN + payload_len])?;
        Ok(Some((Self { msg_type, payload }, HEADER_LEN + payload_len)))
    }

    pub fn handshake(name: &str, screen_width: i32, screen_height: i32) -> Self {
        Self {
            msg_type: MessageType::Handshake,
            payload: serde_json::json!({
                "name": name,
                "screen_width": screen_width,
                "screen_height": screen_height,
            }),
        }
    }

    pub fn handshake_ack(position: &str, server_width: i32, server_height: i32) -> Self {
        Self {
            msg_type: MessageType::HandshakeAck,
            payload: serde_json::json!({
                "position": position,
                "server_width": server_width,
                "server_height": server_height,
            }),
        }
    }

    pub fn key_event(code: u16, value: i32) -> Self {
        Self {
            msg_type: MessageType::KeyEvent,
            payload: serde_json::json!({"code": code, "value": value}),
        }
    }

    pub fn mouse_move(x: i32, y: i32, dx: i32, dy: i32) -> Self {
        Self {
            msg_type: MessageType::MouseMove,
            payload: serde_json::json!({"x": x, "y": y, "dx": dx, "dy": dy}),
        }
    }

    pub fn mouse_button(button: u16, value: i32) -> Self {
        Self {
            msg_type: MessageType::MouseButton,
            payload: serde_json::json!({"button": button, "value": value}),
        }
    }

    pub fn mouse_scroll(dx: i32, dy: i32) -> Self {
        Self {
            msg_type: MessageType::MouseScroll,
            payload: serde_json::json!({"dx": dx, "dy": dy}),
        }
    }

    pub fn switch_to_client(entry_x: i32, entry_y: i32) -> Self {
        Self {
            msg_type: MessageType::SwitchToClient,
            payload: serde_json::json!({"entry_x": entry_x, "entry_y": entry_y}),
        }
    }

    pub fn switch_to_server() -> Self {
        Self {
            msg_type: MessageType::SwitchToServer,
            payload: serde_json::json!({}),
        }
    }

    pub fn heartbeat() -> Self {
        Self {
            msg_type: MessageType::Heartbeat,
            payload: serde_json::json!({}),
        }
    }

    pub fn clipboard_data(content: &str) -> Self {
        Self {
            msg_type: MessageType::ClipboardData,
            payload: serde_json::json!({"content": content}),
        }
    }
}

/// Reads length-framed [`Message`]s off the encrypted [`SecureReader`].
pub struct ProtocolReader {
    reader: SecureReader,
    buffer: Vec<u8>,
}

impl ProtocolReader {
    pub fn new(reader: SecureReader) -> Self {
        Self {
            reader,
            buffer: Vec::new(),
        }
    }

    pub async fn receive(&mut self) -> anyhow::Result<Option<Message>> {
        loop {
            if let Some((msg, consumed)) = Message::decode(&self.buffer)? {
                self.buffer.drain(..consumed);
                return Ok(Some(msg));
            }
            match self.reader.read_chunk().await? {
                Some(plaintext) => self.buffer.extend_from_slice(&plaintext),
                None => return Ok(None),
            }
        }
    }

    /// Drain any additional complete messages already decrypted into the buffer,
    /// without performing another network read.
    pub fn receive_all_buffered(&mut self) -> Vec<Message> {
        let mut messages = Vec::new();
        while let Ok(Some((msg, consumed))) = Message::decode(&self.buffer) {
            self.buffer.drain(..consumed);
            messages.push(msg);
        }
        messages
    }
}

/// Writes length-framed [`Message`]s into the encrypted [`SecureWriter`].
pub struct ProtocolWriter {
    writer: SecureWriter,
}

impl ProtocolWriter {
    pub fn new(writer: SecureWriter) -> Self {
        Self { writer }
    }

    pub async fn send(&mut self, msg: &Message) -> anyhow::Result<()> {
        let data = msg.encode();
        self.writer.write_all(&data).await?;
        Ok(())
    }

    pub async fn flush(&mut self) -> anyhow::Result<()> {
        self.writer.flush().await?;
        Ok(())
    }
}
